flopy/mf6/utils/codegen/filters.py

```diff
@@ -0,0 +1,568 @@
+            if is_array or is_composite:
+                def _args():
+                    args = [
+                        f"'{component_name[1]}'",
+                        f"'{var_block}'",
+                        f"'{var_name}'",
+                    ]
+                    if component_name[0] is not None and component_name[0] not in [
+                        "sim",
+                        "sln",
+                        "utl",
+                        "exg",
+                    ]:
+                        args.insert(0, f"'{component_name[0]}6'")
+                    return args
+
+                kind = "array" if is_array else "list"
+                return f"{var_name} = {kind.title()}TemplateGenerator(({', '.join(_args())}))"
+
+            return None
+
+        attrs = list(filter(None, [_attr(v) for v in component_vars.values()]))
+
+        dfn_file_name = Filters.dfn_file_name(component_name)
+        dfn_header = ["header"]
+        if dfn.get("multi", None):
+            dfn_header.append("multi-package")
+        if dfn.get("advanced", None):
+            dfn_header.append("package-type advanced-stress-package")
+        if dfn.get("sln", None):
+            dfn_header.append(["solution_package", "*"])
+
+        dfn_dir = Path(__file__).parents[2] / "data" / "dfn"
+
+        def _dfn(definition, metadata) -> List[List[str]]:
+            def _meta():
+                exclude = ["subpackage", "parent_name_type"]
+                return [
+                    v for v in metadata if not any(p in v for p in exclude)
+                ]
+
+            def __dfn():
+                def _var(var: dict) -> List[str]:
+                    exclude = ["longname", "description"]
+                    name = var["name"]
+                    subpkg = dfn.get("fkeys", dict()).get(name, None)
+                    if subpkg:
+                        var["construct_package"] = subpkg["abbr"]
+                        var["construct_data"] = subpkg["val"]
+                        var["parameter_name"] = subpkg["param"]
+                    return [
+                        " ".join([k, v]).strip()
+                        for k, v in var.items()
+                        if k not in exclude
+                    ]
+
+                return [_var(var) for var in list(definition.values(multi=True))]
+
+            return [["header"] + _meta()] + __dfn()
+        
+        def _filter_metadata(metadata):
+            meta_ = list()
+            for m in metadata:
+                if "multi" in m:
+                    meta_.append(m)
+                elif "solution" in m:
+                    s = m.split()
+                    meta_.append([s[0], s[2]])
+                elif "package-type" in m:
+                    s = m.split()
+                    meta_.append(" ".join(s))
+            return meta_
+
+        with open(dfn_dir / "common.dfn") as common_f, \
+                open(dfn_dir / dfn_file_name) as f:
```

Comment:
This particular abomination is a temporary way to stick the legacy dfn into the generated class. As soon as we make some progress on mfstructure it can/should go.

(The generated classes all have a [`.dfn` attribute](https://github.com/modflowpy/flopy/blob/3680b7f337c41e80aa518b9c8980719c93d182cf/flopy/mf6/modflow/mfgwfdis.py#L105) currently, which is more or less a verbatim, unstructured copy of the dfn file. `mfstructure.py` then parses this at runtime to do introspection. `mfstructure.py` has proven challenging to unravel.)